From e7b1386e0c4dfec44c918cc872cb3be5c1f74c33 Mon Sep 17 00:00:00 2001
From: David Edmundson <kde@davidedmundson.co.uk>
Date: Wed, 26 Nov 2025 10:34:11 +0000
Subject: [PATCH] inputmethod: Improve keysym lookup from input methods

The ibus wayland bridge for some reason forwards unhandled keys back as
keysyms. The current code has very limited handling of keys including
not being able to type space correctly.

This replaces it with existing code that does full keycode lookup and
can handle anything. Including keysysm outside the range of the current
keycodes.
---
 src/inputmethod.cpp | 71 +++++++++++++++++----------------------------
 src/inputmethod.h   |  1 +
 2 files changed, 28 insertions(+), 44 deletions(-)

diff --git a/src/inputmethod.cpp b/src/inputmethod.cpp
index 4a206790e5..a60ade67e3 100644
--- a/src/inputmethod.cpp
+++ b/src/inputmethod.cpp
@@ -518,26 +518,6 @@ void InputMethod::setEnabled(bool enabled)
     kwinApp()->config()->sync();
 }
 
-static quint32 keysymToKeycode(quint32 sym)
-{
-    switch (sym) {
-    case XKB_KEY_BackSpace:
-        return KEY_BACKSPACE;
-    case XKB_KEY_Return:
-        return KEY_ENTER;
-    case XKB_KEY_Left:
-        return KEY_LEFT;
-    case XKB_KEY_Right:
-        return KEY_RIGHT;
-    case XKB_KEY_Up:
-        return KEY_UP;
-    case XKB_KEY_Down:
-        return KEY_DOWN;
-    default:
-        return KEY_UNKNOWN;
-    }
-}
-
 void InputMethod::keysymReceived(quint32 serial, quint32 time, quint32 sym, bool pressed, quint32 modifiers)
 {
     if (auto t1 = waylandServer()->seat()->textInputV1(); t1 && t1->isEnabled()) {
@@ -560,18 +540,16 @@ void InputMethod::keysymReceived(quint32 serial, quint32 time, quint32 sym, bool
     }
 
     if (effects && effects->hasKeyboardGrab()) {
-        const int keyCode = keysymToKeycode(sym);
-        forwardKeyToEffects(pressed, keyCode, sym);
+        std::optional<Xkb::KeyCode> keyCode = input()->keyboard()->xkb()->keycodeFromKeysym(sym);
+        forwardKeyToEffects(pressed, keyCode.value_or(Xkb::KeyCode{}).keyCode, sym);
         return;
     }
 
-    KeyboardKeyState state;
     if (pressed) {
-        state = KeyboardKeyState::Pressed;
-    } else {
-        state = KeyboardKeyState::Released;
+        forwardKeySym(sym);
+        // reset any modifiers to the actual state
+        input()->keyboard()->xkb()->forwardModifiers();
     }
-    waylandServer()->seat()->notifyKeyboardKey(keysymToKeycode(sym), state);
 }
 
 void InputMethod::commitString(qint32 serial, const QString &text)
@@ -604,23 +582,7 @@ void InputMethod::commitString(qint32 serial, const QString &text)
 
         // First, send all the extracted keys as pressed keys to the client.
         for (const xkb_keysym_t &keySym : keySyms) {
-            std::optional<Xkb::KeyCode> keyCode = input()->keyboard()->xkb()->keycodeFromKeysym(keySym);
-            if (!keyCode) {
-                qCWarning(KWIN_VIRTUALKEYBOARD) << "Could not map keysym " << keySym << "to keycode. Trying custom keymap";
-                static const uint unmappedKeyCode = 247;
-                auto temporaryKeymap = input()->keyboard()->xkb()->keymapContentsForKeysym(unmappedKeyCode, keySym);
-                if (temporaryKeymap.isEmpty()) {
-                    continue;
-                }
-                waylandServer()->seat()->keyboard()->setKeymap(temporaryKeymap);
-                waylandServer()->seat()->notifyKeyboardKey(unmappedKeyCode, KeyboardKeyState::Pressed);
-                waylandServer()->seat()->notifyKeyboardKey(unmappedKeyCode, KeyboardKeyState::Released);
-                waylandServer()->seat()->keyboard()->setKeymap(input()->keyboard()->xkb()->keymapContents());
-            } else {
-                waylandServer()->seat()->notifyKeyboardModifiers(keyCode->modifiers, 0, 0, input()->keyboard()->xkb()->currentLayout());
-                waylandServer()->seat()->notifyKeyboardKey(keyCode->keyCode, KeyboardKeyState::Pressed);
-                waylandServer()->seat()->notifyKeyboardKey(keyCode->keyCode, KeyboardKeyState::Released);
-            }
+            forwardKeySym(keySym);
         }
 
         // reset any modifiers to the actual state
@@ -628,6 +590,27 @@ void InputMethod::commitString(qint32 serial, const QString &text)
     }
 }
 
+void InputMethod::forwardKeySym(int keySym)
+{
+    std::optional<Xkb::KeyCode> keyCode = input()->keyboard()->xkb()->keycodeFromKeysym(keySym);
+    if (!keyCode) {
+        qCWarning(KWIN_VIRTUALKEYBOARD) << "Could not map keysym " << keySym << "to keycode. Trying custom keymap";
+        static const uint unmappedKeyCode = 247;
+        auto temporaryKeymap = input()->keyboard()->xkb()->keymapContentsForKeysym(unmappedKeyCode, keySym);
+        if (temporaryKeymap.isEmpty()) {
+            return;
+        }
+        waylandServer()->seat()->keyboard()->setKeymap(temporaryKeymap);
+        waylandServer()->seat()->notifyKeyboardKey(unmappedKeyCode, KeyboardKeyState::Pressed);
+        waylandServer()->seat()->notifyKeyboardKey(unmappedKeyCode, KeyboardKeyState::Released);
+        waylandServer()->seat()->keyboard()->setKeymap(input()->keyboard()->xkb()->keymapContents());
+    } else {
+        waylandServer()->seat()->notifyKeyboardModifiers(keyCode->modifiers , 0, 0, input()->keyboard()->xkb()->currentLayout());
+        waylandServer()->seat()->notifyKeyboardKey(keyCode->keyCode, KeyboardKeyState::Pressed);
+        waylandServer()->seat()->notifyKeyboardKey(keyCode->keyCode, KeyboardKeyState::Released);
+    }
+}
+
 void InputMethod::deleteSurroundingText(int32_t index, uint32_t length)
 {
     // zwp_input_method_v1 Delete surrounding text interface is designed for text-input-v1.
diff --git a/src/inputmethod.h b/src/inputmethod.h
index 6566ec616c..beb6c1277e 100644
--- a/src/inputmethod.h
+++ b/src/inputmethod.h
@@ -129,6 +129,7 @@ private:
     void resetPendingPreedit();
     void refreshActive();
     void forwardKeyToEffects(bool pressed, int keyCode, int keySym);
+    void forwardKeySym(int keySym);
 
     // buffered till the preedit text is set
     struct
-- 
2.51.2

